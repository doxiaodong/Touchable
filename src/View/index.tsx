import * as React from 'react'
import convertRef from '../convertRef'
import useResponderEvents from '../useResponderEvents'

const View: React.FC<any> = (props) => {
  const {
    // for useResponderEvent
    onMoveShouldSetResponder,
    onMoveShouldSetResponderCapture,
    onResponderEnd,
    onResponderGrant,
    onResponderMove,
    onResponderReject,
    onResponderRelease,
    onResponderStart,
    onResponderTerminate,
    onResponderTerminationRequest,
    onScrollShouldSetResponder,
    onScrollShouldSetResponderCapture,
    onSelectionChangeShouldSetResponder,
    onSelectionChangeShouldSetResponderCapture,
    onStartShouldSetResponder,
    onStartShouldSetResponderCapture,

    // do not copy children
    children,

    // rest for real dom
    ...rest
  } = props

  const child = React.Children.only(props.children)

  const { hostRef, ref } = convertRef(child.ref)

  useResponderEvents(hostRef, {
    onMoveShouldSetResponder,
    onMoveShouldSetResponderCapture,
    onResponderEnd,
    onResponderGrant,
    onResponderMove,
    onResponderReject,
    onResponderRelease,
    onResponderStart,
    onResponderTerminate,
    onResponderTerminationRequest,
    onScrollShouldSetResponder,
    onScrollShouldSetResponderCapture,
    onSelectionChangeShouldSetResponder,
    onSelectionChangeShouldSetResponderCapture,
    onStartShouldSetResponder,
    onStartShouldSetResponderCapture,
  })

  return React.cloneElement(child, {
    ...rest,
    ref,
  })
}

export default View
