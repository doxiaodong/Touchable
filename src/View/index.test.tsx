import * as React from 'react'
import { mount } from 'enzyme'
import toJSON from 'enzyme-to-json'
import View from '.'

// Also test convertRef on next case

describe('View', () => {
  it('snapshot', () => {
    const component = mount(
      <View data-a="aaa">
        <div className="222" />
      </View>
    )

    expect(toJSON(component)).toMatchSnapshot()
  })

  it('pass dom props', () => {
    const component = mount(
      <View className="111">
        <div ref={React.createRef()} />
      </View>
    )

    expect(component.find('div').prop('className')).toBe('111')
  })

  it('View > child', () => {
    const component = mount(
      <View className="111">
        <div className="222" ref={() => null} />
      </View>
    )

    expect(component.find('div').prop('className')).toBe('111')

    // 删除测试
    component.unmount()
  })
})
