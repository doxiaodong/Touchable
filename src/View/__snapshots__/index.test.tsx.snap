// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View snapshot 1`] = `
<View
  data-a="aaa"
>
  <div
    className="222"
    data-a="aaa"
  />
</View>
`;
